File: core/src/main/java/fr/sncf/osrd/api/pathfinding/constraints/LoadingGaugeConstraints.java
package fr.sncf.osrd.api.pathfinding.constraints;

import static fr.sncf.osrd.api.pathfinding.PathfindingUtils.makePath;

import fr.sncf.osrd.sim_infra.api.BlockInfra;
import fr.sncf.osrd.sim_infra.api.Path;
import fr.sncf.osrd.sim_infra.api.RawSignalingInfra;
import fr.sncf.osrd.train.RollingStock;
import fr.sncf.osrd.utils.graph.Pathfinding;
import fr.sncf.osrd.utils.graph.functional_interfaces.EdgeToRanges;
import java.util.Collection;
import java.util.HashSet;
import java.util.stream.Collectors;

public record LoadingGaugeConstraints(
        BlockInfra blockInfra,
        RawSignalingInfra infra,
        Collection<RollingStock> rollingStocks
) implements EdgeToRanges<Integer> {

    @Override
    public Collection<Pathfinding.Range> apply(Integer blockIdx) {
        var res = new HashSet<Pathfinding.Range>();
        var path = makePath(blockInfra, infra, blockIdx);
        for (var stock : rollingStocks)
            res.addAll(getBlockedRanges(stock, path));
        return res;
    }

    /**
     * Returns the sections of the given block that can't be used by the given rolling stock
     */
    private Collection<Pathfinding.Range> getBlockedRanges(RollingStock stock, Path path) {
        return path.getLoadingGauge().asList().stream()
                .filter(entry -> !entry.getValue().isCompatibleWith(stock.loadingGaugeType.ordinal()))
                .map(entry -> new Pathfinding.Range(entry.getLower(), entry.getUpper()))
                .collect(Collectors.toSet());
    }
}

File: core/src/main/kotlin/fr/sncf/osrd/api/pathfinding/constraints/LoadingGaugeConstraints.kt
package fr.sncf.osrd.api.pathfinding.constraints

import fr.sncf.osrd.api.pathfinding.makePathProps
import fr.sncf.osrd.graph.Pathfinding
import fr.sncf.osrd.graph.PathfindingConstraint
import fr.sncf.osrd.railjson.schema.rollingstock.RJSLoadingGaugeType
import fr.sncf.osrd.sim_infra.api.*
import fr.sncf.osrd.utils.DistanceRangeMap
import fr.sncf.osrd.utils.units.Offset
import java.util.stream.Collectors

data class LoadingGaugeConstraints(
    val blockInfra: BlockInfra,
    val infra: RawSignalingInfra,
    val loadingGaugeType: RJSLoadingGaugeType
) : PathfindingConstraint<Block> {
    override fun apply(edge: BlockId): Collection<Pathfinding.Range<Block>> {
        val res = HashSet<Pathfinding.Range<Block>>()
        val path = makePathProps(blockInfra, infra, edge)
        res.addAll(getBlockedRanges(loadingGaugeType, path))
        return res
    }

    /** Returns the sections of the given block that can't be used by the given rolling stock */
    private fun getBlockedRanges(
        type: RJSLoadingGaugeType,
        path: PathProperties
    ): Collection<Pathfinding.Range<Block>> {
        return path
            .getLoadingGauge()
            .asList()
            .stream()
            .filter { (_, _, value): DistanceRangeMap.RangeMapEntry<LoadingGaugeConstraint> ->
                !value.isCompatibleWith(LoadingGaugeTypeId(type.ordinal.toUInt()))
            }
            .map { (lower, upper): DistanceRangeMap.RangeMapEntry<LoadingGaugeConstraint> ->
                Pathfinding.Range(Offset<Block>(lower), Offset(upper))
            }
            .collect(Collectors.toSet())
    }
}
